### PR TITLE
Added the extract job: BigQuery -> Cloud Storage

### PR DESCRIPTION
**Pull request by realAkhmed**
Hello. Have been using this function for a while in my personal fork and always wanted to share it with the community.

This is the code for doing a BigQuery extract job. 
Basically, taking a BigQuery table and extracting it into a CSV file(s) inside Cloud Storage bucket.

The following code contains an example of how to use it

``` r
 # specify your project ID here
project <- "<my_project_id>"

 # specify your Cloud Storage bucket name (note the wildcard)
bucket <- "gs://<my_bucket>/shakespeare*.csv"

# Now run the extract_exec - it will return the number of files that were extracted
extract_exec("publicdata:samples.shakespeare", project = project, destinationUris = bucket)
```

Shakespeare is a small dataset so you won't get charged a lot for that example.

The structure of this file is modeled after `insert_upload_job` and `insert_query_job`

### CODE REVIEW

File: R/extract.r [hunk 1]
```diff
@@ -0,0 +1,89 @@
+#' Create a new extract job.
+#'
+#' This is a low-level function that creates an extract job. To wait until it is
+#' finished, see \code{\link{extract_exec}}
+#'
+#' @param source_table Source table,
+#'   either as a string in the format used by BigQuery, or as a list with
+#'   \code{project_id}, \code{dataset_id}, and \code{table_id} entries
+#' @param project project name
+#' @param destinationUris Specify the extract destination URI. Note: for large files, you may need to specify a wild-card since
```
Comment (craigcitro) [inline R/extract.r]: I think you lost the end of the sentence?

File: R/extract.r [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        printHeader = printHeader
+      )
+    )
+  )
+
+  bq_post(url, body)
+}
+
+
+#' Run a asynchronous extract job and wait till it is done
```
Comment (craigcitro) [inline R/extract.r]: nit: . at end of sentence.

File: R/extract.r [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+#' @export
+#' @examples
+#' \dontrun{
+#' project <- "<my_project_id>" # specify your project ID here
+#' bucket <- "gs://<my_bucket>/shakespeare*.csv" # specify your Cloud Storage bucket name (note the wildcard)
+#'
+#' # Now run the extract_exec - it will return the number of files that were extracted
+#' extract_exec("publicdata:samples.shakespeare", project = project, destinationUris = bucket)
+#' }
+extract_exec <- function(source_table, project, destinationUris,
```
Comment (craigcitro) [inline R/extract.r]: `extract_exec` is a little weird -- maybe just `extract`? (I know it mirrors `query_exec`, but I also think that name is weird. 😉 )

File: R/extract.r [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                               compression = "NONE", destinationFormat = "CSV",
+                               fieldDelimiter = ",", printHeader = TRUE) {
+  assert_that(is.string(source_table), is.string(project))
+
+  job <- insert_extract_job(source_table, project, destinationUris,
+                               compression, destinationFormat,
+                               fieldDelimiter, printHeader)
+  job <- wait_for(job)
+
+  if(job$status$state == "DONE") {
```
Comment (craigcitro) [inline R/extract.r]: Actually, you still need to check for errors, right?

**Comment by hadley** [comment]
@realAkhmed are you interested in finishing off this PR?

**Comment by realAkhmed** [comment]
@hadley Absolutely! Just wasn't sure if the package is still actively developed.

This particular PR was very useful for me internally since it opens way for what I called as `collect_by_export()` -- collecting huge results from BigQuery by automatically saving a query to a temporary table, then exporting it to CSV, then downloading CSV, then parsing CSV locally. This works **much** faster than a regular `collect()` on very large query results.

**Comment by craigcitro** [comment]
@realAkhmed agreed that this is indeed a faster route in general -- but CSV as a transport format isn't great for the case that your table includes nested or repeated fields.

I'd be curious what @hadley knows about potentially converting avro to a dataframe, since that would give us full fidelity exports.

File: R/extract.r [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    (job$statistics$extract$destinationUriFileCounts)
```
Comment (hadley) [inline R/extract.r]: Remove extra parens?

**Comment by IronistM** [comment]
:100: claps for this!

**Comment by hadley** [comment]
I'm just going to merge this so I can work on locally.

**Comment by realAkhmed** [comment]
Thanks @hadley ! Somewhat slow to respond at the moment (the teaching season) -- will come back to address the issues raised by you and @craigcitro once done with it!

**Pull request merged**